biochatter/api_agent/api_agent.py

```diff
@@ -0,0 +1,151 @@
+from typing import Optional
+from collections.abc import Callable
+
+from pydantic import BaseModel
+
+from .abc import BaseFetcher, BaseInterpreter, BaseQueryBuilder
+
+### logic
+
+# 1. User asks question req. API to find the answer
+# 2. rag_agent receives the question and sends it to the api_agent // api_agent is manually selected
+# 3. api_agent writes query for API specific to the question
+# 3.1 question + API prompt template + BlastQuery are input into
+# langchain.chains.openai_functions.create_structured_output_runnable -> returns a structured output which is our API call object
+# 4. API call
+# 4.1 API call is made using the structured output from 3.1
+# 4.2 API returns a response which is saved
+#
+# 5. Read response from and uses it to answer question
+# 6. answer is returned to rag_agent
+
+
+## Agent class
+class APIAgent:
+    def __init__(
+        self,
+        conversation_factory: Callable,
+        query_builder: "BaseQueryBuilder",
+        result_fetcher: "BaseFetcher",
+        result_interpreter: "BaseInterpreter",
+    ):
+        """
+
+        API agent class to interact with a tool's API for querying and fetching
+        results.  The query fields have to be defined in a Pydantic model
+        (`BaseModel`) and used (i.e., parameterised by the LLM) in the query
+        builder. Specific API agents are defined in submodules of this directory
+        (`api_agent`). The agent's logic is implemented in the `execute` method.
+
+        Attributes:
+            conversation_factory (Callable): A function used to create a
+                BioChatter conversation, providing LLM access.
+
+            query_builder (BaseQueryBuilder): An instance of a child of the
+                BaseQueryBuilder class.
+
+            result_fetcher (BaseFetcher): An instance of a child of the
+                BaseFetcher class.
+
+            result_interpreter (BaseInterpreter): An instance of a child of the
+                BaseInterpreter class.
+        """
+        self.conversation_factory = conversation_factory
+        self.query_builder = query_builder
+        self.result_fetcher = result_fetcher
+        self.result_interpreter = result_interpreter
+        self.final_answer = None
+
+    def parameterise_query(self, question: str) -> Optional[BaseModel]:
+        """
+        Use LLM to parameterise a query (a Pydantic model) based on the given
+        question using a BioChatter conversation instance.
+        """
+        try:
+            conversation = self.conversation_factory()
+            return self.query_builder.parameterise_query(question, conversation)
+        except Exception as e:
+            print(f"Error generating query: {e}")
+            return None
+
+    def submit_query(self, api_fields: BaseModel) -> Optional[str]:
```

Comment:
@Tatan47 do we need to rearrange the logic in the APIAgent following the recent changes to returning instead of writing?